hail/scripts/benchmark_in_batch.py

```diff
@@ -0,0 +1,308 @@
+    j.depends_on(*deps)
+    j.always_copy_output()
+
+    for varname, val in env.items():
+        if val is not None:
+            j.env(varname, val)
+
+    j.env('BENCHMARK_TRIAL_ID', str(trial))
+
+    # If the benchmarks fail, we always want this file to exist otherwise
+    # later combine jobs will fail when localising.
+    j.command(f'touch {j.ofile}')
+    j.command('mkdir -p benchmark-resources')
+    j.command(
+        ' '.join([
+            f'python3 -m pytest {path}',
+            '-Werror:::hail -Werror:::hailtop -Werror::ResourceWarning',
+            '--log-cli-level=ERROR',
+            '-s',
+            '-r A',
+            '-vv',
+            '--instafail',
+            '--durations=50',
+            '--timeout=1800',
+            # pytest keeps 3 test sessions worth of temp files by default.
+            # some benchmarks generate very large output files which can quickly
+            # fill the tmpfs and so we'll make pytest always delete tmp files
+            # immediately when tmp_path fixtures are torn-down.
+            '--override-ini=tmp_path_retention_count=0',
+            '--override-ini=tmp_path_retention_policy=failed',
+            f'--output={j.ofile}',
+            '--data-dir=benchmark-resources',
+            f'--iterations={iterations}' if iterations is not None else '',
+            f'-k {benchmark_name}',
+        ])
+    )
+
+    return j.ofile
+
+
+def read_jsonl(p: Path):
+    logging.info(f'reading json lines from {p}.')
+    with p.open(encoding='utf-8') as r:
+        for line in r:
+            if len(line) > 1:
+                yield json.loads(line)
+
+
+def combine(output: Resource, files: List[Resource]) -> None:
+    init_logging()
+    n_files = len(files)
+    if n_files < 1:
+        raise ValueError("'combine' requires at least 1 file to merge")
+
+    logging.info(f'combining {len(files)} files')
+
+    jsonl = [line for f in files for line in read_jsonl(Path(f))]
+    jsonl.sort(key=lambda r: (r['path'], r['name'], r['trial']))
+
+    logging.info(f'Writing combine output to {output}')
+    logging.info(f'collected {len(jsonl)} benchmark jobs.')
+    with open(output, encoding='utf-8', mode='w+') as out:
+        for line in jsonl:
+            json.dump(line, out)
+            out.write('\n')
+
+
+def run_combine(args: Namespace) -> None:
+    combine(args.output, args.files)
+
+
+def make_combine_job(b: Batch, context: str, files: List[Resource]) -> Resource:
+    j = b.new_python_job(f'combine_output_{context}')
+    j.call(combine, j.ofile, files)
+    j.always_run()
+    return j.ofile
+
+
+def combine_results_as_tree(b: Batch, branch_factor: int, results: List[Resource]) -> Resource:
+    phase_i = 1
+
+    while len(results) > 1:
+        results = [
+            (make_combine_job(b, f'phase{phase_i}_job{job_id}', files) if len(files) > 1 else files[0])
+            for files, job_id in zip(chunk(branch_factor, results), count(1))
```

Comment:
because python makes sense and is simple and there's only one way to do things.
```suggestion
            for job_id, files in enumerate(chunk(branch_factor, results), start=1)
```
Will allow deletion of the `from itertools import count` line too.